### ISSUE TITLE
running the testsuite after a rebuild fails with inconsistent assumptions

### ISSUE THREAD

**Issue by gasche** [ocaml/ocaml#11970]
I have been hacking on top of my `trunk` branch for a while (full build, running the testsuite etc.). Then a few days later I update my `trunk`, rebase my branch on top, and this pulls a configuration change (the reserved header bits stuff). So I run:

```
$ ./configure
$ make -j
```

which works well, and then I run

```
$ make -C testsuite/ parallel
[...]
make[1]: Entering directory '/home/gasche/Prog/ocaml/github-trunk/testsuite/tools'
File "_none_", line 1:
Error: Files expect_test.cmo and ../../compilerlibs/ocamlcommon.cma(Warnings)
       make inconsistent assumptions over interface Warnings
make[1]: *** [Makefile:60: expect_test] Error 2
```

I have seen similar failures many many times in the past, and I have a routine to work around it (`make -C testsuite clean && make -C testsuite/tools expect_test`), but this disrupts my work and I wish I would not have to do this.

**Comment by gasche** [comment]
Note: when I say "this introduces a configuration change, so I (configure and make again)", what happens in practice is that when I `make -j` again, I see this (this is only the very end of the error messages, which are 822 lines long):

```
runtime/meta.c: In function ‘caml_realloc_global’:
runtime/caml/mlvalues.h:123:51: error: ‘HEADER_RESERVED_BITS’ undeclared (first use in this function); did you mean ‘HEADER_WOSIZE_BITS’?
  123 |                             - HEADER_COLOR_BITS - HEADER_RESERVED_BITS)
      |                                                   ^~~~~~~~~~~~~~~~~~~~
runtime/caml/mlvalues.h:125:39: note: in expansion of macro ‘HEADER_WOSIZE_BITS’
  125 | #define HEADER_WOSIZE_MASK (((1ull << HEADER_WOSIZE_BITS) - 1ull) \
      |                                       ^~~~~~~~~~~~~~~~~~
runtime/caml/mlvalues.h:129:44: note: in expansion of macro ‘HEADER_WOSIZE_MASK’
  129 | #define Wosize_hd(hd) ((mlsize_t) (((hd) & HEADER_WOSIZE_MASK) \
      |                                            ^~~~~~~~~~~~~~~~~~
runtime/caml/mlvalues.h:172:26: note: in expansion of macro ‘Wosize_hd’
  172 | #define Wosize_val(val) (Wosize_hd (Hd_val (val)))
      |                          ^~~~~~~~~
runtime/meta.c:164:17: note: in expansion of macro ‘Wosize_val’
  164 |   actual_size = Wosize_val(old_global_data);
      |                 ^~~~~~~~~~
make: *** [Makefile:960: runtime/meta.b.o] Error 1
```

Now I happen to follow OCaml development closely enough to know that this means "you need to run `./configure` again", but it might be possible to say this in a clearer way.

**Comment by dra27** [comment]
Looking only at the second failure, and it may now be hard to answer, but are you certain that you'd re-run configure at this stage? The recent changes altered `runtime/caml/m.h.in` to add `HEADER_RESERVED_BITS` and re-running `configure` should have caused `runtime/caml/m.h` to be regenerated? That error _looks_ like you ran `git pull ; make -j` without `configure`?

If one wanted this fully automated, files such as `runtime/caml/m.h` should depend on `runtime/caml/m.h.in` _and_ `config.status` (which itself should depend on `configure`) and either automatically re-run them or at least warn/fail the build. I'm for that kind of alteration, but it adds complexity which has been objected to in the past (parts in the old `utils/Makefile`, etc.).

I think the testsuite issue can be fairly simply fixed by making all the .cmo files produced by the tools depend on the compiler-libs which are linked... they're cheap to compile, so there'd be no issue just rebuilding them whenever ocamlcommon.cma changes.

**Comment by gasche** [comment]
> Looking only at the second failure, and it may now be hard to answer, but are you certain that you'd re-run configure at this stage?

Apologies for being unclear. What I meant is that I observed this output, *before* re-configuring, and that I deduced from the output that I should re-configure.

> If one wanted this fully automated, files such as runtime/caml/m.h should depend on runtime/caml/m.h.in and config.status (which itself should depend on configure) and either automatically re-run them or at least warn/fail the build. I'm for that kind of alteration, but it adds complexity which has been objected to in the past (parts in the old utils/Makefile, etc.).

Yes, I understand the trade-off here. Maybe another option would be to have logic in place *after* the configure stage to check that the configuration values are valid (in the expected range of outputs produced by configure), and fail with a clean error (which would suggest to reconfigure) when they are not. This could be, for example, static assertions included in the configure-generated .h files.

**Comment by dra27** [comment]
Just to aid coming back to this. The testsuite failure can be seen by rebuilding across 5f05c2b0cf and 0ec8679c337 (which is the merge of PR#11635, which changed the interface of `Warnings`). The configuration failure is seen by instead building f1550145d4 (which is the merge of PR#11872, which introduced `RESERVED_HEADER_BITS`). Both can be seen with this Dockerfile:

```Dockerfile
RUN git clone --shallow-since=2023-01-01 https://github.com/ocaml/ocaml.git
WORKDIR ocaml
RUN git checkout 5f05c2b && ./configure && make -j && make -C testsuite tools

FROM base-5f05c2b AS base-0ec8679
RUN git checkout 0ec8679
RUN make -j

# Re-configure failure can be seen with merge of #11871
FROM base-5f05c2b
RUN git checkout f155014
RUN make -j || true

# Testsuite rebuild failure can be seen with merge of #11635
FROM base-0ec8679
RUN make -C testsuite all
```

**Comment by dra27** [comment]
Possible bandaid for the testsuite:

https://github.com/dra27/ocaml/commit/29f6c0c6b40154caff7f0e484231b4c86063cf17: that would over-rebuild, but it's simpler than making, say, expect and codegen objects depend on _just_ the compiler-libs needed by each tool. Technically, I don't think having `.cmi` depend on `.cma` catches absolutely all scenarios, but it does fix the above.

Less bandaid-y, for the re-configuration warning:

https://github.com/dra27/ocaml/commit/d284f39d2ad9088d79a83874ec6197de90f7acc3: this commit would only help once it's incorporated (i.e. it's no good if the tree you originally built is _before_ this commit, so bisecting wouldn't ever benefit), but in the Dockerfile example above you now get:
```
Makefile:1867: *** You need to re-run configure.  Stop.
Changed configuration file(s): Makefile.build_config.in Makefile.config.in utils/config.generated.ml.in runtime/caml/m.h.in
```

**Comment by gasche** [comment]
Thanks for the work in this direction! Minor remarks in case it helps:
- The "configuration issues could warn properly" is secondary I think (... or really I should have created a separate issue)
- I see absolutely no issue with rebuilding the testsuite tools "too much"; they could be built on each run of the testsuite as far as I'm concerned, as they are very fast to build.

**Comment by gasche** [comment]
If I understand correctly, both issues reported have been fixed in #11987 and #11988. Thanks @dra27 and @shindere! Closing.

**Comment by dra27** [comment]
You're welcome! Pleasingly brief solutions to both in the end.

**Comment by dra27** [comment]
#11988 doesn't work with header files, as `config.status` doesn't write header files unless they've actually changed. Re-opening this just to track that part of it.

**Comment by dra27** [comment]
I'd anticipate this has been addressed by #12706, as the tools are now built as part of the main build (in advance of running the testsuite). If it's not, please re-open!
